src/core/qgsactionmanager.cpp

```diff
@@ -56,21 +57,68 @@ QUuid QgsActionManager::addAction( QgsAction::ActionType type, const QString &na
 
 void QgsActionManager::addAction( const QgsAction &action )
 {
+  QgsDebugMsg( "add action " + action.name() );
   mActions.append( action );
+  if ( mLayer && mLayer->dataProvider() && !action.notificationMessage().isEmpty() )
+  {
+    mLayer->dataProvider()->setListening( true );
+    if ( !mOnNotifyConnected )
+    {
+      QgsDebugMsg( "connecting to notify" );
+      connect( mLayer->dataProvider(), &QgsDataProvider::notify, this, &QgsActionManager::onNotifyRunActions );
+      mOnNotifyConnected = true;
+    }
+  }
+}
+
+void QgsActionManager::onNotifyRunActions( const QString &message )
+{
+  for ( const QgsAction &act : qgsAsConst( mActions ) )
```

Comment:
Better iterate on a copy of this `QgsAction`, while it's totally unlikely we never know if the container is not modified inside `runAction` and we don't want to risk a crash.

Reply:
we pass a copy of the action to runAction, so we are safe, or I'm missing something.